### ISSUE TITLE
How to Run PAS Operations in Parallel?

### ISSUE THREAD

**Issue by brmcconn** [pspete/psPAS#220]
**Describe the issue**
This isn't so much of an issue, but more of a general question I wanted to throw out here. I am attempting to setup a jenkins pipeline that would leverage this module to interact with the CyberArk API. This module works great to accomplish everything I need to do. 

However, this pipeline would be used by multiple people and could potentially have parallel executions. I was planning/hoping to setup the pipeline to authenticate with the API using the same account on each execution. Given the concept of initiating a session for the user, if someone kicks off a pipeline execution while one is already executing, that new execution would initiate a new session, effectively expiring/invalidating the session on the prior/in progress execution causing any of the functions executing to return:

Invoke-PASRestMethod : [401] The session token is missing, invalid or expired.

Is there any way around this beyond having the pipeline use a different account to authenticate for each execution such as the specific user account of the user executing the pipeline?

**To Reproduce**
1. Execute New-PASSession to initiate a session in powershell.
2. Execute New-PASSession to initiate a session in another powesehll execution.
3. In the initial powershell session, attempt to run something like Get-PASAccount, which will now error with "Invoke-PASRestMethod : [401] The session token is missing, invalid or expired."

```

**Your Environment**
Include relevant details about your environment

* PowerShell Version: 5.1.17763.771
* psPAS Version: 3.2.37
* CyberArk Version: 10.7.0.32

**Comment by pspete** [comment]
Hey @brmcconn , 

Two immediately available options to consider:

The older "Classic API" authentication endpoint accepts a `-connectionNumber` parameter, allowing multiple sessions to be opened by the same account:

```powershell
New-PASSession -UseClassicAPI -connectionNumber $x -Credential $cred -BaseURI $url
New-PASSession -UseClassicAPI -connectionNumber $y -Credential $cred -BaseURI $url
New-PASSession -UseClassicAPI -connectionNumber $z -Credential $cred -BaseURI $url
```

It is also possible to configure/use [Shared Logon Authentication](https://docs.cyberark.com/Product-Doc/OnlineHelp/PAS/Latest/en/Content/SDK/REST%20Web%20Services%20API%20-%20Shared%20Logon%20Authentication.htm), which allows multiple connections using a shared user. Strong recommendation to also pair this with certificate authentication:
```powershell
New-PASSession -UseSharedAuthentication -BaseURI $url
New-PASSession -UseSharedAuthentication -BaseURI $url-CertificateThumbprint $CertificateThumbprint
```
There are other approaches too, but all depends on what works best for you and your implementation.

**Comment by brmcconn** [comment]
@pspete - Thanks! This looks promising. I gave both options a try. Unfortunately, both are failing:

![image](https://user-images.githubusercontent.com/21693427/69888607-610de280-12b2-11ea-8214-3a894678e195.png)

For the classic API option, I passed my credentials I've been using successfully for other tests into $cred using $cred =  Get-Credential. I'm guessing that either doesn't work on the version we're using (CyberArk Version: 10.7.0.32) or it's something that needs to be enabled. I'll reach out to our vault admin(s) to discuss what I'm trying to do and see if they're willing to enable shared auth.

**Comment by pspete** [comment]
Shared Logon Auth does require additional configuration before it can be used. See the link in my previous response, it is all detailed there.
I tested the ClassicAPI authentication method against an 11.1 system and saw no issues; work with you vault admin to understand the reason for the 403 response and why your credentials do not work against the Classic API.

**Comment by AaronG1234** [comment]
i am having some success running in parallel with
##Login
New-Passession -baseurl ..........................

then
$Session = Get-PASSession
$ThreadCode = {param($thissession,$safe); Use-PASSession $thissession;Get-PASAccount -filter "Safename eq $safe"}

Start-ThreadJob -ScriptBlock $Scriptblock -Arg $session,"XXXXWORKSTATIONS"
Start-ThreadJob -ScriptBlock $Scriptblock -Arg $session,"XXXMACOS"

The jobs run in parrelel and both finish with proper data (both these requests return thousands of accounts)

Aaron

**Comment by AaronG1234** [comment]
Last Comment on this ill leave on this... 
For my hardware, my PVWA and my Vault Architecture... 5 parallel tasks from the same Powershell client, using the same Authenticated user and sharing a websession is optimal.  i ran some very intensive tests and 6,7,8 threads completed in the same amount of time as 5.  2,3,4, were progressively faster (i had about 20 jobs of 350 Account Reads and 10-15 Account Changes per each job,, then throttled to 5 jobs ever running at one time)

(Note the sharing works only with Threadjobs, regular PS Jobs did not work, the PASSession got serialized and totally messed up) ... 

Also The tasks must be more expensive than about 1-2 minutes or the creation of the thread itself was too expensive.  In my particular use case each thread took about 40 additional secs compared to just running the chuck complete solo, however my job had a very big scriptblock, so that is probably not a big factor in many use cases.

**Comment by brmcconn** [comment]
Sorry for my radio silence here. I've been side tracked with some other things, but I did reach out to the team that manages our password vault implementation and ran the issues I'm having by them. They discovered that the ConnectionNumber parameter was removed in V10 and submitted a ticket to CyberArk inquiring what the alternative is here with Windows Auth being the required authentication method for us. I'm going to go ahead and close this out as it seems like more of an issue I'll have to work out on my end. Thanks again for the responses!
